### DISCUSSION TITLE
How to use IfcConstraint

### DISCUSSION THREAD

**Discussion by nickger** [IfcOpenShell/IfcOpenShell#4394] [Q&A]
Does anyone know how to use these IfcConstraint objects, Relations etc.? Can it be developed to a "true" parametric IFC?

**Comment by theoryshaw** [comment]
Curious too...
https://community.osarch.org/discussion/comment/11538/#Comment_11538

**Comment by Moult** [comment]
There are two types of constraints: user defined and system defined.

The user defined ones are currently "restricted" to only constrain properties or quantities such as in the diagram on the IfcMetric docs. They are not necessarily enforced, and arguably superseded by IDS.

The system defined ones are more advanced and can in theory affect anything, but are defined in the IFC docs themselves. For example IfcConstructionResource has a constraint association allow you to lock and balance resource utilisation, time duration, and cost.

Note that IFC is an ISO standard. IFC serves the international common denominator of BIM data and is therefore unlikely to define parametric rules other than "standard case" scenarios. That said, we can always build on top of it.

**Comment by nickger** [comment]
Well, that's exactly what I am asking: how to use these constraints. 
IFC documentation is usless as usual, the link to GeometryGym looks better. I know, many years ago there was Constructivity software which claimed to be able to build parametric IFC objects, but I haven't seen its source code or explanation, how it is compliant with IFC schema.
@Moult  As you said, these constraints are not supposed to be purely geometrical, but arbitrary. Can we make a "self-aware" IFC? With built-in rules as constraints? 
Can we write cheking rules in IFC format and prove other files against them?
